File: src/DelegateClassTest.java
/**
 * Created by Whyte on 2017/7/10.
 */
public class DelegateClassTest {
    public static void main(String[] args) {
        L2 c = new L2();
        c.f();//L
        c.toL1();
        c.f();//L1
        c.toL();
        c.f();//L
    }
}

interface I {
    void f();
}

class L implements I {
    public void f() {
        System.out.println("L");
    }
}

class L1 implements I {
    public void f() {
        System.out.println("L1");
    }
}

class L2 implements I {
    //将L对象委托为I接口，再通过I去调用
    private I i = new L();

    public void f() {
        i.f();
    }

    /**
     * 新建L对象
     */
    public void toL() {
        i = new L();
    }

    /**
     * 新建L1对象
     */
    public void toL1() {
        i = new L1();
    }
}

File: src/DelegateClassTest.kt
/**
 * Created by Whyte on 2017/7/10.
 */
fun main(args: Array<String>) {
    P3(P1()).print()//输出 p1
    P3(P2()).print()//输出 p2
}

interface P {
    fun print()
}

class P1 : P {
    override fun print() {
        println("p1")
    }
}

class P2 : P {
    override fun print() {
        println("p2")
    }
}

//超类型列表中的by子句表示b将会在 P3 中内部存储，并且编译器将生成转发给b的所有P的方法。
class P3(b: P) : P by b



